File: maze/Maze.cpp

#include "Maze.h"

//Room* _rooms[MAZE_SIZE];

Maze::Maze() {
    _currentRoom = 0;
}

void Maze::AddRoom(Room* room) {
    if (_currentRoom < MAZE_SIZE) {
        _rooms[_currentRoom++] = room;
    }
}

Room* Maze::roomNo(int roomNumber) const {
    if (roomNumber < MAZE_SIZE) {
        return _rooms[roomNumber];
    } else {
        return NULL;
    }
}

void Maze::printStatus() {
    //rooms
    cout << "Maze contains " << _currentRoom  << " rooms" << endl;

    for(int i = 0; i < _currentRoom; i++) {
        _rooms[i]->printStatus();
    }
}

File: FactoryMethod/src/com/ash/maze/Maze.java
package com.ash.maze;

import java.util.ArrayList;

public class Maze {
    private final ArrayList<Room> rooms;
    private int currentRoom;

    public Maze() {
        this.rooms = new ArrayList<>();
        this.currentRoom = 0;
    }

    public void addRoom(Room room) {
        this.rooms.add(this.currentRoom++, room);
    }

    public Room getRoom(int roomNumber) {
        if (roomNumber < this.rooms.size()) {
            return this.rooms.get(roomNumber);
        } else {
            return null;
        }
    }

    @Override
    public String toString() {
        StringBuffer description = new StringBuffer();

        for (int i = 0; i < this.rooms.size(); i++) {
            description.append(this.rooms.get(i) + "\n");
        }

        return description.toString();
    }
}
